new/missing-method-missing.md

```diff
@@ -0,0 +1,213 @@
+
+    As a Chef developer,
+    I want meaningful names in stack traces for methods,
+    So that I can actually read the debug output.
+
+    As a Chef developer,
+    I want less if statements, case statements and metaprogramming in DSL method calls,
+    So that I can debug and follow logic more easily.
+
+## Specification
+
+### Create explicit DSL methods for resources
+
+When looking up resources, there is no method anywhere that corresponds to the
+actual resource name.  This can lead to a lot of unnecessarily heavy thinking
+when following some very typical stack traces.  It also means that the name of
+the resource being declared is often not on the stack, making debugging harder.
+
+Here we propose that all resource and definition DSL be added to
+`Chef::DSL::Recipe` as actual methods, as they are created, so that they can be
+subject to inspection and the stack traces can be followed.  This will happen
+automatically when the user declares resources and definitions.
+
+### Automatically add new resources and providers to DSL, regardless of namespace
+
+Presently, when you type `my_resource` in the Chef DSL, in the absence of any
+`provides` directives to direct Chef to the correct resource class, the
+namespaces `Chef::Resource` is explicitly searched for
+`Chef::Resource::MyResource`.
+
+This means that if you have seen examples of classes in Chef::Resource, and want
+to do the obvious thing and just move to your own namespace, it doesn't
+immediately work and it is far from apparent why. The result? A massive traffic
+jam of non-core resources in the `Chef::Resource` namespace and puzzling error
+messages for cookbook conflicts such as "superclass mismatch for class MyResource."
+
+Chef will now automatically call `provides` on all new, named resource classes,
+using `ResourceClass.dsl_name` (a transform of the class name, i.e.
+MyPackage::MyResource -> my_resource) as the provided class.
+
+```ruby
+class MyPackage::MyResource < Chef::Resource
+end
+# now you can type 'my_resource' in a recipe.
+```
+
+#### Explicit vs. Implicit `provides`
+
+If you specify `provides :something_else` in a class, that will override the
+auto-class-creation behavior.
+
+```ruby
+class MyPackage::MyResource < Chef::Resource
+  provides :shazam
+end
+# Now `shazam` will work fine in a recipe, but `my_resource` will not.
+```
+
+If you want to still provide the original name, you must be explicit:
+
+```ruby
+class MyPackage::MyResource < Chef::Resource
+  provides :my_resource
+  provides :shazam
+end
+# Now both `shazam` and `my_resource` work in a recipe
+```
+
+#### `does_not_provide`
+
+If you want to (for example) create a base class that does not have an actual
+DSL method, you can specify `does_not_provide` or `does_not_provide :my_name` in
+the class:
+
+```ruby
+class MyPackage::MyResourceBase < Chef::Resource
+  does_not_provide
+end
+# my_resource_base will not exist in recipe DSL
+```
```

Comment:
Deprecating the implicit magic would cause a large amount of backcompat woes.  I'd like that to be a separate RFC.

My feeling is, Chef::Resource should always do implicit magic for backcompat, and when we move to some sort of real base mixin that is our true public API for resource, we can just not implement implicitness there.